### PR TITLE
Lock all leaves for insert on root of AO partition table.

### PR DESCRIPTION
**Pull request by kainwen**
**NOTES for reviewers:**

1. this is to 6X, 6X does not save `lockmode` in RTE, so we need to handle in different places (like for cached plan)
2. previous commit 99fbc8ea9496a9c61e759ae2134c4961b3ae59fc introduces the logic to fix for a concurrent issue, please refer to its test cases for details.
3. I also include the global deadlock test case here.
4. **MOST IMPORTANT**: we have concern on the lock mode of leaf, refer to previously discussion: https://github.com/greenplum-db/gpdb/pull/13538#discussion_r891830143, greenplum's AO table does not support concurrently update|delete, but if we want to lock leaf partitions with polymorphic storage, the API `find_all_inheritors`  only support a specific lockmode as argument. This PR does not change behavior, we might fix this later?

--------

Previously when GDD is enabled, Greenplum will not lock leaf
partitions for insert statement on root of partition table no mattter
what storage type is.

The above locking leafs logic is added by commit 99fbc8ea, please
refer to it for more details and here we only providea simple
backgroup knowledge on the global deadlock issue without locking
leafs:
  1. txn A: insert into root_part;
  6. txn B: truncate leaf1;
  7. txn A and B will not block each other on QD if no locking leaf
  8. suppose on seg0 txn A arrive first, seg1 txn B arrive first
  9. on seg0, txn B will wait for A
  10. on seg1, txn A will wait for B
  11. Global Deadlock happens

That is why when GDD is disabled, previous code will lock all leaf
partitions even for insert statement on root. When GDD is enabled, we
think even deadlock happens, GDD can break it thus skip locking all
leaf partition for insert on root. NOTE: global deadlock might also
happen due to AO|AOCO segfile information update.

Skip locking for insert statement on AOCO partition's root might lead
to some other problem, even with GDD is enabled. Take a look at
ATExecSetDistributedBy function, at the end of it, it will try to
remove the entry in AppendOnly Hash table (shared hash table), that
code assume no one will touch it because ExclusiveLock is held for
Alter Table statement (Refer to the commments of that function). A
simple case to demonstrate is:
  1. txn A: insert into AO partition's root (with GDD enabled)
  2. txn A just finish its InitPlan() on QD
  3. txn B: alter table leaf1 set with(reorg=true), removes the hash
            table entry in shared hash table
  4. Global Deadlock happen, txn B is terminated by GDD
  6. txn A continue to update segfile information, and cannot find the
  entry in the shared hash table.

This commit fixes the issue by also locking leafs for insert statement
directly on AO|AOCO's root partition even GDD is enabled.

Also remove the test case planner_insert_while_vacuum_drop because now
we lock all leafs for AO partition table.

### CODE REVIEW

File: src/backend/parser/parse_clause.c [hunk 1]
```diff
@@ -342,51 +344,62 @@ setTargetTable(ParseState *pstate, RangeVar *relation,
...
-			lockmode = NoLock;
+		if(pstate->p_is_insert && gp_enable_global_deadlock_detector && !is_ao)
+		{
+			/*
+			 * Do nothing in this block.
+			 *
+			 * We skip locking leaf partitions for DML on root partition when
+			 * the following 3 conditions are all satisfied:
+			 *
+			 *   1. insert statement directly on the root partitition
```
Comment (zxuejing) [inline src/backend/parser/parse_clause.c:362]: partitition-->partition

File: src/backend/parser/parse_clause.c [hunk 2]
```diff
@@ -342,51 +344,62 @@ setTargetTable(ParseState *pstate, RangeVar *relation,
...
+			 *   2. GDD is enabled
+			 *   3. the target relation is heap
+			 *
+			 * Why condition 3 in the above? Greenplum 6X will update information
+			 * of AO|AOCS tables' segment files on QD, and this can only be done
+			 * at the end of the executor (mppExecutorFinishup) and needs to lock
+			 * the leaf partition at that time. Locking a relation first time at
+			 * the end of executition is not good, this might lead to global
```
Comment (zxuejing) [inline src/backend/parser/parse_clause.c:370]: executition -->execution

File: src/test/isolation2/sql/gdd/insert_root_partition_truncate_deadlock.sql [hunk 3]
```diff
@@ -0,0 +1,83 @@
+-- Insert statement on root partition, if GDD is enabled, QD will not lock leaf
+-- parititions for better concurrency performance. So when the insert statement
```
Comment (zxuejing) [inline src/test/isolation2/sql/gdd/insert_root_partition_truncate_deadlock.sql:2]: partititions-->partition

**Review by zxuejing — APPROVED** [review]

**Comment by ashwinstar** [comment]
> MOST IMPORTANT: we have concern on the lock mode of leaf, refer to previously discussion: (6x only) Lock all leaves for insert on root of AO partition table. #13538 (comment), greenplum's AO table does not support concurrently update|delete, but if we want to lock leaf partitions with polymorphic storage, the API find_all_inheritors only support a specific lockmode as argument. This PR does not change behavior, we might fix this later?

I don't feel we can push this to fix later. Like making any kind of decision-based on root partitions storage type is incorrect and hence this PR will yield unintended behavior if the root is a heap and all children are AO or vice-versa, no? I think we will have to go with a conservative approach and lock every partition if it's a partitioned table irrespective of its heap/ao. Unless we add code to check the full hierarchy is heap table and apply a different locking strategy.

**Comment by kainwen** [comment]
> > MOST IMPORTANT: we have concern on the lock mode of leaf, refer to previously discussion: (6x only) Lock all leaves for insert on root of AO partition table. #13538 (comment), greenplum's AO table does not support concurrently update|delete, but if we want to lock leaf partitions with polymorphic storage, the API find_all_inheritors only support a specific lockmode as argument. This PR does not change behavior, we might fix this later?
> 
> I don't feel we can push this to fix later. Like making any kind of decision-based on root partitions storage type is incorrect and hence this PR will yield unintended behavior if the root is a heap and all children are AO or vice-versa, no? I think we will have to go with a conservative approach and lock every partition if it's a partitioned table irrespective of its heap/ao. Unless we add code to check the full hierarchy is heap table and apply a different locking strategy.

Then seems we need to firstly loop and lock all leafs using AccessShareLock, and fetch each leaf's storage type, and then lock each leaf using the correct lock mode, and then unlock the AccessShareLock. Seems make the code rather ugly.

**Comment by ashwinstar** [comment]
> Then seems we need to firstly loop and lock all leafs using AccessShareLock, and fetch each leaf's storage type, and then lock each leaf using the correct lock mode, and then unlock the AccessShareLock. Seems make the code rather ugly.

- Lets first establish common understanding - making any kind of decision-based on root partitions storage type will yield unintended behavior if the root is a heap and all children are AO or vice-versa, right? Please can you write test to prove or disprove this
- once we are on common page and know its problem we can evaluate solutions (ugly or not :)) If we wish to avoid ugliness then can just have a conservative approach and lock every partition if it's a partitioned table irrespective of its heap/ao.

**Comment by kainwen** [comment]
@ashwinstar Thanks for your comments.

> Lets first establish common understanding - making any kind of decision-based on root partitions storage type will yield unintended behavior if the root is a heap and all children are AO or vice-versa, right? Please can you write test to prove or disprove this

Sure. This will lead to some problem.

---------

So now I decide to fix the two issues in two commits. (**NOTE**:  master branch also has the issue, because it use the root partition's RTE's lockmode, the same logic as is now for 6X). 

For master, I'd wait for some time to see if we can support concurrent DML for AO tables so that storage type will not impact lock mode anymore, if so the issue is gone. 

Now let's focus on my proposal for fixing this in 6X.

**Commit 1**: fix lockmode for DELETE|UPDATE on a mixed storage type partition table.

1. I will introduce a new API like `find_all_inheritors`, but will take storage type into consideration. For example, when gdd is enabled, a root partition is heap, and leaf1 is AO, leaf2 is AO, leaf3 is heap; then a DELETE statement will lock the root, leaf3 using RowExclusive, lock leaf1, leaf2 using Exclusive.
2. if GDD is not enabled, lock all leafs in ExclusiveMode.

**Commit 2**: fix insert for partition table will lock all leafs.

1. always lock all leafs using RowExclusive.
  - if GDD is not enabled, this is needed to avoid global deadlock
  - if GDD is enabled, this is to fix AO's shared hash entry race condition. Since we might have mixed storage types, it is not good to lock AO leaf and not-lock heap-leaf. Let's just lock all leaf in RowExclusive always.


---------

Any ideas? Thanks.

**Comment by ashwinstar** [comment]
> For master, I'd wait for some time to see if we can support concurrent DML for AO tables so that storage type will not impact lock mode anymore, if so the issue is gone.

Sure. Based on AO visimap design though I feel its not possible, we can continue to explore. What could be possible on master branch is with GDD let Segments do exclusive locking only and not coordinator. Though that may just open up for many deadlocks which will get resolved by GDD. With taking lock on master avoids that deadlock possibility today.

For GPDB6:
- Given we are coming up with new API based on storage type curious to learn why don't make its use consistent for insert/update/delete instead of locking all partitions for insert vs selectively for update/delete?
- this new API seems would be much heavy compared to current one. As current one just scans though pg_inherits table. New one will have to read and join with pg_class to find the object type
- Ignoring partitions so currently with GDD enabled INSERTS don't lock for heap tables? and DELETE/UPDATE takes RowExclusive lock? why is that difference if exists
- seems we need to continue to think through on this topic and explore more alternative options

**Comment by kainwen** [comment]
Close this. Need more dicussion.

**Pull request closed without merging**